Implementing throttle axis accumulator with deadband support

gregd72002 opened the pull request:
Ref: https://github.com/mavlink/qgroundcontrol/issues/4117

This relates only to Joysticks/Gamepads

- adds option for throttle accumulator
- deadband support on all axis when in accumulator mode (requires re-calibration)

rjehangir commented:
@gregd72002 Are you actually using the deadband? It looks like it's hardcoded to zero?

I've been working on a deadband for the joystick code as well so I'd like to coordinate efforts.

gregd72002 commented:
@rjehangir sure, my deadband implementation is very rudimental. It was just implemented into to satisfy the accumulator on throttle so it does not change without user action.

In this implementation:
- deadband is set to 0 initially
- during 1st stage of calibration where sticks are supposed to be centred, it probes the deviations to get the highest for each of the axis; this is done in JoystickConfigController.cc (_inputCenterWaitBegin)
- this is then used in Joystick.cc (_adjustRange) 
- this will only work when axis are centred around 0, hence I apply the deadband only when in throttle accumulator mode (what implies 0 centre)

I'm sure there are better way of doing it. What's your take on it?

rjehangir commented:
@gregd72002 Okay. Thanks for the explanation.

The deadband I implemented is really simple. It's just a fixed value around the center. I think the main joystick that needs the deadband is the Xbox controller so I experimentally chose the deadband value to work well with that. Most other controllers that I've tried have very clearly defined deadbands by default. 

Since I'm working mostly with ArduSub, my throttle is also centered around zero.

I added a checkbox to the joystick page to turn the deadband on/off.

Thoughts? Yours is a little more sophisticated as long as it works reliably.

-Rusty

gregd72002 commented:
It would be nice to be able to specify custom deadband as opposed to 'sensed' or 'pre-defined'.

Possibly it should be specified as % as opposed to axis units - Thoughts?
Maybe introduce new fields next to 'Axis Monitor' that allows user to define deadband for each axis. I could program it but QML is not my strength...

I have been testing this with PS3 controller.

rjehangir commented:
@gregd72002 I personally think that adding the ability to specify a custom deadband is too much control to the user and would probably end up confusing them. I'd prefer to keep things as simple and clear as possible. I imagine we could take care of 95% of joysticks with a fixed deadband width.

Thoughts?

gregd72002 commented:
From my experience joystick do change deadband due to wear and tear. While hardcoding the deadband would work for me maybe the sensing option is more versatile?

I put together a quick mockup:
- the deadband is now an option
- it is sensed as described before
- when active, the deadbands are shown red on each of the axis

<img width="1300" alt="screen shot 2016-10-12 at 23 55 38" src="https://cloud.githubusercontent.com/assets/161846/19330759/3c04a050-90d8-11e6-8204-da2c4d676e6f.png">

DonLakeFlyer commented:
Deadband:
* I'm with Rusty. Too many options and you scare people off. 
* The auto-detect is nice, but I would also include a fixed deadband value. Use the max of both.
* I would also warn if auto-detect deadband is too high. This could cause mysterious behavior to the end user.
* Any reason why deadband shouldn't be on all the time? Fewer options the better.

Accumulator:
* If like to find a different name for this. Maybe something alongs the line of "smoothing"?

rjehangir commented:
@DonLakeFlyer If a fixed deadband is always applied then it will effectively increase the deadband for joysticks that already have large deadbands, like the Logitech 310. Might not be a huge deal though.

DonLakeFlyer commented:
Ok, you guys know way more about this than I do. Just trying to remove an option for simplification.

gregd72002 commented:
I've added the warning message and renamed a couple of things.

What do you think about the labels? 

<img width="1298" alt="screen shot 2016-10-13 at 23 31 10" src="https://cloud.githubusercontent.com/assets/161846/19369420/28a48ef6-919d-11e6-8bf4-3ffa60a6f8f3.png">

rjehangir commented:
@gregd72002 What if we just limit the auto-detected deadband to 0-10% so it can never be set to a crazy number? Then we could get rid of the warning.

Also, as a side note, with the switch to SDL2, I'm hoping to eventually eliminate the calibration process entirely. SDL2 has consistent stick and button numbering so there is no need to detect axes anymore. The deadband could not be detected from that, unfortunately.

Last, is "Spring on throttle support" the new name for accumulator?

gregd72002 commented:
@rjehangir I do not see particular reason for the warning message. The axis clearly show the deadband that is being applied. Also, deadbands are only in use when user explicitly activates them. The implication in here is that they know what deadbands are before activating them. 
For all the accidental deadbands we have the red marker on the axis.

Re SDL2, once the calibration is removed we could introduce user defined deadbands?

"Spring on throttle support" is the new name I proposed for the accumulator.

DonLakeFlyer commented on src/Joystick/Joystick.cc:
```diff
@@ -109,9 +116,13 @@ void Joystick::_loadSettings(void)
 
-        qCDebug(JoystickLog) << "_loadSettings axis:min:max:trim:reversed:badsettings" << axis << calibration->min << calibration->max << calibration->center << calibration->reversed << badSettings;
+
+        qCDebug(JoystickLog) << "_loadSettings axis:min:max:trim:reversed:badsettings" << axis << calibration->min << calibration->max << calibration->center << calibration->reversed << calibration->deadband << badSettings;
```
Can you update the text for the new values?

DonLakeFlyer commented on src/Joystick/Joystick.cc:
```diff
@@ -161,14 +175,16 @@ void Joystick::_saveSettings(void)
         settings.setValue(revTpl.arg(axis), calibration->reversed);
+        settings.setValue(deadbndTpl.arg(axis), calibration->deadband);
 
         qCDebug(JoystickLog) << "_saveSettings name:axis:min:max:trim:reversed"
```
Same text update here?

DonLakeFlyer commented on src/Joystick/Joystick.cc:
```diff
@@ -214,7 +238,7 @@ float Joystick::_adjustRange(int value, Calibration_t calibration)
                             << calibration.max
                             << calibration.center
-                            << calibration.center
+                            << calibration.deadband
```
Remove center? Shouldn't this be like the others with an add of deadband? Also not sure if there is text update needed here.

gregd72002 replied:
center is printed 2 times, I replaced one of them

DonLakeFlyer commented on src/Joystick/Joystick.h:
```diff
@@ -66,6 +67,8 @@ class Joystick : public QThread
 
     Q_PROPERTY(int throttleMode READ throttleMode WRITE setThrottleMode NOTIFY throttleModeChanged)
     Q_PROPERTY(int exponential READ exponential WRITE setExponential NOTIFY exponentialChanged)
```
Shouldn't all of these be bool (except throttle mode)?

gregd72002 replied:
updated.

DonLakeFlyer commented on src/VehicleSetup/JoystickConfig.qml:
```diff
@@ -87,6 +87,20 @@ SetupPage {
                         color:                  __barColor
                     }
 
+                    // Deadband
```
It's sort of interesting to display this, but I really wonder whether it will just confuse users as to what it is?

gregd72002 replied:
By default deadbands will be disabled as such there should be no difference to users. It is only when they active them.

DonLakeFlyer replied:
Sound fine then.

DonLakeFlyer commented on src/VehicleSetup/JoystickConfig.qml:
```diff
@@ -382,6 +407,36 @@ SetupPage {
+                                    QGCCheckBox {
+                                        id:         deadband
+                                        checked:    _activeJoystick.deadband
+                                        text:       qsTr("Deadband (requires re-calibration)")
```
I'd make this stupid-proof and invalid the calibration if deadband is turned on. Either that or always calc deadband and save so it's always there, but just don't use if not turned on.

gregd72002 commented:
PR refreshed with the comments.

DonLakeFlyer commented:
With the headband display I can live without the warning. I still don't like the accumulator naming though. I think you want to describe what it does. My logitech F430 (name?) seems to work fine with it's spring loaded throttle. So I wouldn't use this. But if you call it just "spring loaded throttle" then it would push people to use this when they may not need to. Maybe add "smoothing" to the end up of so you hit both: "Spring loaded throttle smoothing"?

gregd72002 commented:
Done

DonLakeFlyer commented:
Ok, this is good from review. I'm going to do some testing with it then I'll merge.

DonLakeFlyer commented:
Ran this today. Shouldn't deadbands be supported for throttle down zero as well?

gregd72002 commented:
Well, this is not what I was trying to achieve. But I guess I get your point. Should we:
- move the deadband toggle to Advanced Options on the Joystick Setup (always available)
- when throttle centre 0 mode - the behaviour is as per current implementation
- when throttle down 0 mode - this deadband is not applied for throttle but applied for other axis
?

DonLakeFlyer commented:
Yeah, I think that all makes sense.

gregd72002 commented:
This has been updated. Not sure why the check is failing - looking into the details shows some mavlink related stuff which should be irrelevant to this PR?
